bitcoin/src/address/mod.rs

```diff
@@ -308,103 +308,108 @@ pub enum AddressData {
-///
-/// ```
-/// # use bitcoin::address::{Address, NetworkChecked};
-/// let address: Address<NetworkChecked> = "132F25rTsvBdp9JzLLBHP5mvGY66i1xdiM".parse::<Address<_>>()
-///                .unwrap().assume_checked();
-/// assert_eq!(format!("{:?}", address), "132F25rTsvBdp9JzLLBHP5mvGY66i1xdiM");
-/// ```
-///
-/// ### Relevant BIPs
-///
-/// * [BIP13 - Address Format for pay-to-script-hash](https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki)
-/// * [BIP16 - Pay to Script Hash](https://github.com/bitcoin/bips/blob/master/bip-0016.mediawiki)
-/// * [BIP141 - Segregated Witness (Consensus layer)](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki)
-/// * [BIP142 - Address Format for Segregated Witness](https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki)
-/// * [BIP341 - Taproot: SegWit version 1 spending rules](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)
-/// * [BIP350 - Bech32m format for v1+ witness addresses](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-// The `#[repr(transparent)]` attribute is used to guarantee the layout of the `Address` struct. It
-// is an implementation detail and users should not rely on it in their code.
-#[repr(transparent)]
-pub struct Address<V = NetworkChecked>(AddressInner, PhantomData<V>)
-where
-    V: NetworkValidation;
+internals::transparent_newtype! {
+    /// A Bitcoin address.
+    ///
+    /// ### Parsing addresses
+    ///
+    /// When parsing string as an address, one has to pay attention to the network, on which the parsed
+    /// address is supposed to be valid. For the purpose of this validation, `Address` has
+    /// [`is_valid_for_network`](Address<NetworkUnchecked>::is_valid_for_network) method. In order to provide more safety,
+    /// enforced by compiler, `Address` also contains a special marker type, which indicates whether network of the parsed
+    /// address has been checked. This marker type will prevent from calling certain functions unless the network
+    /// verification has been successfully completed.
+    ///
+    /// The result of parsing an address is `Address<NetworkUnchecked>` suggesting that network of the parsed address
+    /// has not yet been verified. To perform this verification, method [`require_network`](Address<NetworkUnchecked>::require_network)
+    /// can be called, providing network on which the address is supposed to be valid. If the verification succeeds,
+    /// `Address<NetworkChecked>` is returned.
+    ///
+    /// The types `Address` and `Address<NetworkChecked>` are synonymous, i. e. they can be used interchangeably.
+    ///
+    /// ```rust
+    /// use std::str::FromStr;
+    /// use bitcoin::{Address, Network};
+    /// use bitcoin::address::{NetworkUnchecked, NetworkChecked};
+    ///
+    /// // variant 1
+    /// let address: Address<NetworkUnchecked> = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf".parse().unwrap();
```

Comment:
This api seems fine, although, why not combine parsing and validation in a single call?

```rust
let address: Address<_> = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf".parse(Network::Bitcoin).unwrap();
```

Reply:
Furthermore, when parsing string to u64, an example might be:

```
let number = some_str.parse::<u64>().unwrap();
```

That also parses and does a check in one go. so the equiv would be:

```
let address: Address<_> = "32iVBEu4dxkUQk9dJbZUiBiQdmypcEyJRf"::<Network::Bitcoin>.unwrap();
```

Reply:
Firstly, this is unrelated, secondly, `parse` comes from `FromStr` which doesn't have the extra parameter so what you're proposing is impossible. Your last example is invalid syntax but anyway, you seem to have intended hard-coding the network which is *very* bad for testability and such thing will never get ACK from me.

Reply:
> you seem to have intended hard-coding the network

I don't follow.  A network address belongs to a network though, so binding a network to an address seems appropriate.  My example is meant more as sudo code but sorry for the invalid syntax mistakes.

Reply:
s/sudo/Pseudo lol

Reply:
The specific network must not be a type parameter. Otherwise it'll motivate people to just hard code it. The way pretty much everyone does it now is they have per-process configuration (reading at the start of the program) and just check all inputs. That's a very sensible way of doing it that doesn't have horrible consequences.

Reply:
Alright.  I wasn't trying to slow down this PR, just commenting out of curiosity.